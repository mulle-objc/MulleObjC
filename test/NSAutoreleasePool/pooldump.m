#ifndef __MULLE_OBJC__
# import <Foundation/Foundation.h>
# pragma message "this test does not work with Apple Foundation"
#else
# import <MulleObjC/MulleObjC.h>
#endif


@interface Foo : NSObject
{
   char   _mulleNameUTF8String[ 32];
}

@property( dynamic, assign) char   *mulleNameUTF8String  MULLE_OBJC_THREADSAFE_PROPERTY;

@end



@implementation NSInvocation( Squelch)

- (char *) mulleNameUTF8String  MULLE_OBJC_THREADSAFE_PROPERTY
{
   return( "unnamed");
}

@end



@implementation Foo

+ (instancetype) objectWithNameUTF8String:(char *) s
{
   Foo  *obj;

   obj = [Foo alloc];
   obj = [obj init];
   strncpy( obj->_mulleNameUTF8String, s ? s : "", sizeof( obj->_mulleNameUTF8String) - 1);
   obj = [obj autorelease];
   return( obj);
}



- (char *) mulleNameUTF8String
{
   if( ! _mulleNameUTF8String[ 0])
      mulle_snprintf( _mulleNameUTF8String, sizeof( _mulleNameUTF8String), "<Foo %p>", self);
   return( &_mulleNameUTF8String[ 0]);
}


- (void) mulleSetNameUTF8String:(char *) s
{
   strncpy( _mulleNameUTF8String, s ? s : "", sizeof( _mulleNameUTF8String) - 1);
}


@end



static int   test( NSThread *thread, id unused)
{
   Foo  *b;

   [[NSAutoreleasePool mulleDefaultAutoreleasePool] mulleSetNameUTF8String:"thread root"];
   b = [Foo objectWithNameUTF8String:"b"];  // place an object there

   @autoreleasepool
   {
      // place b also here again, twice
      [[b retain] autorelease];
      [[b retain] autorelease];

      [[NSAutoreleasePool mulleDefaultAutoreleasePool] mulleSetNameUTF8String:"thread inner"];
      [Foo objectWithNameUTF8String:"c"];  // place an object there

      // now dump all to standard out
      MulleObjCDumpAutoreleasePoolsToFILEWithOptions( stdout, 0x1);
   }
   return( 0);
}



int  main()
{
   NSThread   *thread;
   Class      cls;

   [[NSAutoreleasePool mulleDefaultAutoreleasePool] mulleSetNameUTF8String:"main root"];
   [Foo objectWithNameUTF8String:"a"];  // place an object there

   cls    = [Foo class];
   thread = [NSThread instantiate];
   thread = [thread mulleInitWithObjectFunction:test
                                         object:nil];
   [thread mulleSetNameUTF8String:"test-thread"];
   [thread mulleStart];
   [thread mulleJoin];

   return( 0);
}
