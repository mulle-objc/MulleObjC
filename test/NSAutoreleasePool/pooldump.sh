#! /bin/sh

./pooldump-dot.awk "$1.csv" > "$1.dot" \
&& dot -Tsvg "$1.dot" -o "$1.svg" \
&& xdg-open "$1.svg"

