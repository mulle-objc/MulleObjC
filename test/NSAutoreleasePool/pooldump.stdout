thread_id,thread,pool_id,pool,object_id,object,class,rc
0,"NSMainThread",0,"main root",0,"a","Foo",1
0,"NSMainThread",0,"main root",1,"test-thread","NSThread",2
1,"test-thread",1,"thread inner",2,"b","Foo",3
1,"test-thread",1,"thread inner",2,"b","Foo",3
1,"test-thread",1,"thread inner",3,"c","Foo",1
1,"test-thread",0,"thread root",2,"b","Foo",3
